nano-sync/src/circuits/mnt4/pk_tree_5.rs

```diff
@@ -0,0 +1,314 @@
+            })?;
+
+        let commit_agg_pk_var =
+            G2Gadget::alloc(cs.ns(|| "alloc commit aggregate public key"), || {
+                Ok(&self.commit_agg_pk)
+            })?;
+
+        // Allocate all the public inputs.
+        next_cost_analysis!(cs, cost, || { "Alloc public inputs" });
+
+        let pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc public keys commitment"),
+            self.pks_commitment.as_ref(),
+        )?;
+
+        let prepare_signer_bitmap_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc prepare signer bitmap"),
+            self.prepare_signer_bitmap.as_ref(),
+        )?;
+
+        let prepare_agg_pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc prepare aggregate pk commitment"),
+            self.prepare_agg_pk_commitment.as_ref(),
+        )?;
+
+        let commit_signer_bitmap_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc commit signer bitmap"),
+            self.commit_signer_bitmap.as_ref(),
+        )?;
+
+        let commit_agg_pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc commit aggregate pk commitment"),
+            self.commit_agg_pk_commitment.as_ref(),
+        )?;
+
+        let position_var =
+            UInt8::alloc_input_vec(cs.ns(|| "alloc position"), self.position.as_ref())?
+                .pop()
+                .unwrap();
+
+        // Process public inputs.
+        next_cost_analysis!(cs, cost, || { "Process public inputs" });
+
+        let chunk_start = VALIDATOR_SLOTS / PK_TREE_BREADTH * (self.position[0] as usize);
+        let chunk_end = VALIDATOR_SLOTS / PK_TREE_BREADTH * (1 + self.position[0] as usize);
+
+        let prepare_signer_bitmap_var =
+            prepare_signer_bitmap_var.to_bits(cs.ns(|| "prepare signer bitmap to bits"))?;
+
+        let prepare_signer_bitmap_var = prepare_signer_bitmap_var[chunk_start..chunk_end].to_vec();
+
+        let commit_signer_bitmap_var =
+            commit_signer_bitmap_var.to_bits(cs.ns(|| "commit signer bitmap to bits"))?;
+
+        let commit_signer_bitmap_var = commit_signer_bitmap_var[chunk_start..chunk_end].to_vec();
+
+        let mut path_var = position_var.into_bits_le();
+
+        path_var.truncate(PK_TREE_DEPTH);
+
+        // Verify the Merkle proof for the public keys.
+        next_cost_analysis!(cs, cost, || { "Verify Merkle proof pks" });
+
+        let mut bits: Vec<Boolean> = vec![];
+
+        for i in 0..self.pks.len() {
+            bits.extend(SerializeGadget::serialize_g2(
+                cs.ns(|| format!("serialize pks: {}", i)),
+                &pks_var[i],
+            )?);
+        }
+
+        MerkleTreeGadget::verify(
+            cs.ns(|| "verify Merkle proof for pks"),
+            &bits,
+            &pks_nodes_var,
+            &path_var,
+            &pk_commitment_var,
+            &pedersen_generators_var,
+            &sum_generator_g1_var,
+        )?;
```

Comment:
I would add a comment here to remind people what the merkle tree is composed of.
As far as I understood it contains the concatenated public keys for each chunk as its leaves, right?

Reply:
Not concatenated, aggregated. But yeah, I'll add a comment.

Reply:
Done!